process/process_exit_group.go

```diff
@@ -0,0 +1,156 @@
+package process
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"unsafe"
+
+	ps "github.com/mitchellh/go-ps"
+	"golang.org/x/sys/windows"
+)
+
+// We use this struct to retreive process handle(which is unexported)
+// from os.Process using unsafe operation.
+type handleRetreiver struct {
+	Pid    int
+	Handle uintptr
+}
+
+type processExitGroup windows.Handle
+
+type ProcessTreeRoot struct {
+	Children []*ProcessTreeNode
+}
+
+func (tree *ProcessTreeRoot) AddChild(child *ProcessTreeNode) error {
+	tree.Children = append(tree.Children, child)
+	return nil
+}
+
+func (tree *ProcessTreeRoot) String() string {
+	result := ""
+	for _, childNode := range tree.Children {
+		result += childNode.String(1)
+	}
+	return result
+}
+
+type ProcessTreeNode struct {
+	Process ps.Process
+	Children []*ProcessTreeNode
+}
+
+func (node *ProcessTreeNode) AddChild(child *ProcessTreeNode) error {
+	node.Children = append(node.Children, child)
+	return nil
+}
+
+func (node *ProcessTreeNode) String(depth int) string {
+	result := ""
+	result += strings.Repeat(" ", depth * 2) + fmt.Sprintf("%d - %s\n", node.Process.Pid(), node.Process.Executable())
+	for _, childNode := range node.Children {
+		result += childNode.String(depth + 1)
+	}
+	return result
+}
+
+func newProcessExitGroup() (processExitGroup, error) {
+	handle, err := windows.CreateJobObject(nil, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	// this could be neat for enforcing clean up of all processes, but for
+	// now we're only interested in improved observability
+	//info := windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION{
+	//	BasicLimitInformation: windows.JOBOBJECT_BASIC_LIMIT_INFORMATION{
+	//		LimitFlags: windows.JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE,
+	//	},
+	//}
+	//if _, err := windows.SetInformationJobObject(
+	//	handle,
+	//	windows.JobObjectExtendedLimitInformation,
+	//	uintptr(unsafe.Pointer(&info)),
+	//	uint32(unsafe.Sizeof(info))); err != nil {
+	//	return 0, err
+	//}
```

Comment:
Job objects are super interesting, and it **seems* like they might provide a way for us to reliably terminate **all** processes that run as part of a job. I'm nervous that enabling it now might make it harder to observe the root issue at play in #1292 though, so haven't included them in this PR for now.